internal/services/elastic/elastic.go

```diff
@@ -0,0 +1,257 @@
+func (s *Service) GetHistory(ctx context.Context, params GetHistoryParams) ([]json.RawMessage, error) {
+	if params.Buckets < 1 {
+		return nil, fmt.Errorf("non-positive number of buckets: %w", ErrInvalidParams)
+	}
+	// build the query
+	req := buildHistoryQuery(params)
+
+	// perform the request
+	res, err := s.esClient.Search().Index(s.settings.DeviceDataIndexNameV2).Request(&req).Perform(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to complete request: %w", err)
+	}
+
+	//nolint:errcheck // we don't care about the error closing the body.
+	defer res.Body.Close()
+
+	// check for errors any status code over 299 is an error
+	if res.StatusCode >= http.StatusMultipleChoices {
+		body, _ := io.ReadAll(res.Body) // we ignore the error here because we are already in an error state
+		return nil, fmt.Errorf("got status code %d from Elastic message = %s", res.StatusCode, string(body))
+	}
+
+	body, err := io.ReadAll(res.Body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read response body: %w", err)
+	}
+
+	// depending on the start and end time we may get len(results) <=  buckets + 1
+	// pass the max number of values to getHitsFromHistoryResponse to ensure we only get the desired number of values
+	return getHitsFromHistoryResponse(body, params.Buckets), nil
+}
+
+// SetDefaultHistoryParams sets the default values for the history params.
+// If the buckets are less than or equal to 0, it will be set to 1000.
+// If the end time is zero, it will be set to the current time.
+// If the start time is zero, it will be set to 2 weeks from the end time.
+func (g *GetHistoryParams) SetDefaultHistoryParams() {
+	if g.Buckets <= 0 {
+		g.Buckets = 1000
+	}
+	if g.EndTime.IsZero() {
+		g.EndTime = time.Now()
+	}
+
+	if g.StartTime.IsZero() {
+		g.StartTime = g.EndTime.Add(-time.Hour * 24 * 14) // default to 2 weeks ago
+	}
+}
+
+// buildHistoryQuery builds the query to retrieve the history of a device from elastic.
+func buildHistoryQuery(params GetHistoryParams) search.Request {
+	// convert the start and end times to milliseconds since epoch
+	startArg := strconv.Itoa(int(params.StartTime.UnixMilli()))
+	endArg := strconv.Itoa(int(params.EndTime.UnixMilli()))
+
+	// calculate the interval in milliseconds
+	interval := calculateIntervalMS(params.StartTime, params.EndTime, params.Buckets)
+	intervalArg := strconv.Itoa(interval) + "ms"
+
+	msm := types.MinimumShouldMatch(1)
+	min := 1
+	filter := getHistoricalPrivilegeFilter(params.PrivilegeIDs)
+	return search.Request{
+		// get documents that match the device ID, are within the time range, and have a vehicle field
+		Query: &types.Query{
+			Bool: &types.BoolQuery{
+				Filter: []types.Query{
+					{Term: map[string]types.TermQuery{subjectField: {Value: params.DeviceID}}},
+					{Range: map[string]types.RangeQuery{timeField: types.DateRangeQuery{Gte: &startArg, Lte: &endArg}}},
+				},
+				Should: []types.Query{
+					{Exists: &types.ExistsQuery{Field: "data.vehicle"}},
```

Comment:
@Allyson-English Is there a reason this `exists` thing is here in should and not in the filter?

Reply:
hmm I don't think so actually, looking at it now it seems like it'd make more sense to be moved up.. especially since minimum should make on just one item seems unnecessary